File: src/api-service/__app__/onefuzzlib/endpoint_authorization.py
#!/usr/bin/env python
#
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.

import logging
import urllib
from typing import Callable, Optional
from uuid import UUID

import azure.functions as func
from azure.functions import HttpRequest
from memoization import cached
from onefuzztypes.enums import ErrorCode
from onefuzztypes.models import Error, UserInfo

from .azure.creds import get_scaleset_principal_id
from .azure.group_membership import create_group_membership_checker
from .config import InstanceConfig
from .request import not_ok
from .request_access import RequestAccess
from .user_credentials import parse_jwt_token
from .workers.pools import Pool
from .workers.scalesets import Scaleset


@cached(ttl=60)
def get_rules() -> Optional[RequestAccess]:
    config = InstanceConfig.fetch()
    if config.api_access_rules:
        return RequestAccess.build(config.api_access_rules)
    else:
        return None


def check_access(req: HttpRequest) -> Optional[Error]:
    rules = get_rules()

    # Nothing to enforce if there are no rules.
    if not rules:
        return None

    path = urllib.parse.urlparse(req.url).path
    rule = rules.get_matching_rules(req.method, path)

    # No restriction defined on this endpoint.
    if not rule:
        return None

    member_id = UUID(req.headers["x-ms-client-principal-id"])

    try:
        membership_checker = create_group_membership_checker()
        allowed = membership_checker.is_member(rule.allowed_groups_ids, member_id)
        if not allowed:
            logging.error(
                "unauthorized access: %s is not authorized to access in %s",
                member_id,
                req.url,
            )
            return Error(
                code=ErrorCode.UNAUTHORIZED,
                errors=["not approved to use this endpoint"],
            )
    except Exception as e:
        return Error(
            code=ErrorCode.UNAUTHORIZED,
            errors=["unable to interact with graph", str(e)],
        )

    return None


@cached(ttl=60)
def is_agent(token_data: UserInfo) -> bool:

    if token_data.object_id:
        # backward compatibility case for scalesets deployed before the migration
        # to user assigned managed id
        scalesets = Scaleset.get_by_object_id(token_data.object_id)
        if len(scalesets) > 0:
            return True

        # verify object_id against the user assigned managed identity
        principal_id: UUID = get_scaleset_principal_id()
        return principal_id == token_data.object_id

    if not token_data.application_id:
        return False

    pools = Pool.search(query={"client_id": [token_data.application_id]})
    if len(pools) > 0:
        return True

    return False


def can_modify_config_impl(config: InstanceConfig, user_info: UserInfo) -> bool:
    if config.admins is None:
        return True

    return user_info.object_id in config.admins


def can_modify_config(req: func.HttpRequest, config: InstanceConfig) -> bool:
    user_info = parse_jwt_token(req)
    if not isinstance(user_info, UserInfo):
        return False

    return can_modify_config_impl(config, user_info)


def check_require_admins_impl(
    config: InstanceConfig, user_info: UserInfo
) -> Optional[Error]:
    if not config.require_admin_privileges:
        return None

    if config.admins is None:
        return Error(code=ErrorCode.UNAUTHORIZED, errors=["pool modification disabled"])

    if user_info.object_id in config.admins:
        return None

    return Error(code=ErrorCode.UNAUTHORIZED, errors=["not authorized to manage pools"])


def check_require_admins(req: func.HttpRequest) -> Optional[Error]:
    user_info = parse_jwt_token(req)
    if isinstance(user_info, Error):
        return user_info

    # When there are no admins in the `admins` list, all users are considered
    # admins.  However, `require_admin_privileges` is still useful to protect from
    # mistakes.
    #
    # To make changes while still protecting against accidental changes to
    # pools, do the following:
    #
    # 1. set `require_admin_privileges` to `False`
    # 2. make the change
    # 3. set `require_admin_privileges` to `True`

    config = InstanceConfig.fetch()

    return check_require_admins_impl(config, user_info)


def is_user(token_data: UserInfo) -> bool:
    return not is_agent(token_data)


def reject(req: func.HttpRequest, token: UserInfo) -> func.HttpResponse:
    logging.error(
        "reject token.  url:%s token:%s body:%s",
        repr(req.url),
        repr(token),
        repr(req.get_body()),
    )
    return not_ok(
        Error(code=ErrorCode.UNAUTHORIZED, errors=["Unrecognized agent"]),
        status_code=401,
        context="token verification",
    )


def call_if(
    req: func.HttpRequest,
    method: Callable[[func.HttpRequest], func.HttpResponse],
    *,
    allow_user: bool = False,
    allow_agent: bool = False
) -> func.HttpResponse:

    token = parse_jwt_token(req)
    if isinstance(token, Error):
        return not_ok(token, status_code=401, context="token verification")

    if is_user(token):
        if not allow_user:
            return reject(req, token)

        access = check_access(req)
        if isinstance(access, Error):
            return not_ok(access, status_code=401, context="access control")

    if is_agent(token) and not allow_agent:
        return reject(req, token)

    return method(req)


def call_if_user(
    req: func.HttpRequest, method: Callable[[func.HttpRequest], func.HttpResponse]
) -> func.HttpResponse:

    return call_if(req, method, allow_user=True)


def call_if_agent(
    req: func.HttpRequest, method: Callable[[func.HttpRequest], func.HttpResponse]
) -> func.HttpResponse:

    return call_if(req, method, allow_agent=True)

File: src/ApiService/ApiService/onefuzzlib/EndpointAuthorization.cs
using System.Net.Http;
using Microsoft.Azure.Functions.Worker.Http;
using Microsoft.Extensions.Logging;
using Microsoft.Graph;
namespace Microsoft.OneFuzz.Service;

public record UserAuthInfo(UserInfo UserInfo, List<string> Roles);

public interface IEndpointAuthorization {
    Async.Task<OneFuzzResultVoid> CheckRequireAdmins(UserAuthInfo authInfo);
    Async.Task<(bool, string)> IsAgent(UserAuthInfo authInfo);
    Async.Task<OneFuzzResultVoid> CheckAccess(HttpRequestData req);
}


public class EndpointAuthorization : IEndpointAuthorization {
    private readonly IOnefuzzContext _context;
    private readonly ILogger _log;
    private readonly GraphServiceClient _graphClient;
    private static readonly IReadOnlySet<string> _agentRoles = new HashSet<string>() { "UnmanagedNode", "ManagedNode" };

    public EndpointAuthorization(IOnefuzzContext context, ILogger<EndpointAuthorization> log, GraphServiceClient graphClient) {
        _context = context;
        _log = log;
        _graphClient = graphClient;
    }

    public async Async.Task<OneFuzzResultVoid> CheckRequireAdmins(UserAuthInfo authInfo) {
        var config = await _context.ConfigOperations.Fetch();
        if (config is null) {
            return Error.Create(
                ErrorCode.INVALID_CONFIGURATION,
                "no instance configuration found ");
        }

        return CheckRequireAdminsImpl(config, authInfo.UserInfo);
    }

    private static OneFuzzResultVoid CheckRequireAdminsImpl(InstanceConfig config, UserInfo userInfo) {
        // When there are no admins in the `admins` list, all users are considered
        // admins.  However, `require_admin_privileges` is still useful to protect from
        // mistakes.
        //
        // To make changes while still protecting against accidental changes to
        // pools, do the following:
        //
        // 1. set `require_admin_privileges` to `False`
        // 2. make the change
        // 3. set `require_admin_privileges` to `True`

        if (config.RequireAdminPrivileges == false) {
            return OneFuzzResultVoid.Ok;
        }

        if (config.Admins is null) {
            return Error.Create(ErrorCode.UNAUTHORIZED, "pool modification disabled ");
        }

        if (userInfo.ObjectId is Guid objectId) {
            if (config.Admins.Contains(objectId)) {
                return OneFuzzResultVoid.Ok;
            }

            return Error.Create(ErrorCode.UNAUTHORIZED, "not authorized to manage instance");
        } else {
            return Error.Create(ErrorCode.UNAUTHORIZED, "user had no Object ID");
        }
    }

    public async Async.Task<OneFuzzResultVoid> CheckAccess(HttpRequestData req) {
        var instanceConfig = await _context.ConfigOperations.Fetch();

        var rules = GetRules(instanceConfig);
        if (rules is null) {
            return OneFuzzResultVoid.Ok;
        }

        var path = req.Url.AbsolutePath;
        var rule = rules.GetMatchingRules(new HttpMethod(req.Method), path);
        if (rule is null) {
            return OneFuzzResultVoid.Ok;
        }

        var memberId = Guid.Parse(req.Headers.GetValues("x-ms-client-principal-id").Single());
        try {
            var membershipChecker = CreateGroupMembershipChecker(instanceConfig);
            var allowed = await membershipChecker.IsMember(rule.AllowedGroupsIds, memberId);
            if (!allowed) {
                _log.LogError("unauthorized access: {MemberId} is not authorized to access {Path}", memberId, path);
                return Error.Create(ErrorCode.UNAUTHORIZED, "not approved to use this endpoint");
            } else {
                return OneFuzzResultVoid.Ok;
            }
        } catch (Exception ex) {
            return Error.Create(ErrorCode.UNAUTHORIZED, "unable to interact with graph", ex.Message);
        }
    }

    private GroupMembershipChecker CreateGroupMembershipChecker(InstanceConfig config) {
        if (config.GroupMembership is not null) {
            return new StaticGroupMembership(config.GroupMembership);
        }

        return new AzureADGroupMembership(_graphClient);
    }

    private static RequestAccess? GetRules(InstanceConfig config) {
        var accessRules = config?.ApiAccessRules;
        if (accessRules is not null) {
            return RequestAccess.Build(accessRules);
        }

        return null;
    }

    public async Async.Task<(bool, string)> IsAgent(UserAuthInfo authInfo) {
        if (!_agentRoles.Overlaps(authInfo.Roles)) {
            return (false, "no agent role");
        }

        var tokenData = authInfo.UserInfo;

        if (tokenData.ObjectId != null) {
            var scalesets = _context.ScalesetOperations.GetByObjectId(tokenData.ObjectId.Value);
            if (await scalesets.AnyAsync()) {
                return (true, string.Empty);
            }

            var principalId = await _context.Creds.GetScalesetPrincipalId();
            if (principalId == tokenData.ObjectId) {
                return (true, string.Empty);
            }
        }

        if (!tokenData.ObjectId.HasValue) {
            return (false, "no object id in token");
        }

        var pools = _context.PoolOperations.GetByObjectId(tokenData.ObjectId.Value);
        if (await pools.AnyAsync()) {
            return (true, string.Empty);
        }

        return (false, "no matching scaleset or pool");
    }
}
